components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -272,133 +314,496 @@ do_registration_ends ()
+                                                                             dyn_dims(i,0),
+                                                                             dyn_dims(i,1),
+                                                                             dyn_dims(i,2),
+                                                                             dyn_dims(i,3),
+                                                                             dyn_dims(i,4),
+                                                                             dyn_dims(i,5));
+
+      const size_t N = dyn.extent(0)*dyn.extent(1)*dyn.extent(2)*
+                       dyn.extent(3)*dyn.extent(4)*dyn.extent(5);
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team, N), [&](const int& k) {
+        int k0   = k%dyn.extent(0);
+        int ktmp = k/dyn.extent(0);
+        int k1   = ktmp%dyn.extent(1);
+        ktmp     = ktmp/dyn.extent(1);
+        int k2   = ktmp%dyn.extent(2);
+        ktmp     = ktmp/dyn.extent(2);
+        int k3   = ktmp%dyn.extent(3);
+        ktmp     = ktmp/dyn.extent(3);
+        int k4   = ktmp%dyn.extent(4);
+        int k5   = ktmp/dyn.extent(4);
+        dyn(k0, k1, k2, k3, k4, k5) = 0;
+      });
+      break;
+    }
+    default:
+      ekat::error::runtime_abort("Error! Unexpected layout.\n");
+  }
+}
+
+
+template<typename RealType>
+void PhysicsDynamicsRemapper<RealType>::
+do_remap_fwd() const
+{
+  using KT = KokkosTypes<DefaultDevice>;
 
   const auto& tl = Homme::Context::singleton().get<Homme::TimeLevel>();
 
-  const int num_fields = m_dyn.size();
-  for (int i=0; i<num_fields; ++i) {
-    const auto& phys = m_phys[i];
-    const auto& dyn  = m_dyn[i];
+  const int num_fields = m_phys.size();
+  const int num_cols   = m_phys_grid->get_num_local_dofs();
+  const auto lid2elgp  = m_dyn_grid->get_lid_to_idx_map();
+  const auto p2d       = m_p2d;
 
-    const auto& ph = phys.get_header();
-    const auto& dh = dyn.get_header();
+  const auto field_loop = KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const int i = team.league_rank();
 
-    const auto& phys_alloc_prop = ph.get_alloc_properties();
-    const auto& dyn_alloc_prop  = dh.get_alloc_properties();
+    switch (phys_layout(i)) {
+      case 0:
+      case 1:
+      case 2:
+      {
+        if (pack_alloc_property(i) == 0) {
+          set_dyn_to_zero<pack_type>(team);
+          team.team_barrier();
 
-    const bool phys_pack_alloc = phys_alloc_prop.template is_compatible<pack_type>();
-    const bool dyn_pack_alloc  = dyn_alloc_prop.template  is_compatible<pack_type>();
-    const bool phys_small_pack_alloc = phys_alloc_prop.template is_compatible<small_pack_type>();
-    const bool dyn_small_pack_alloc  = dyn_alloc_prop.template  is_compatible<small_pack_type>();
+          local_remap_fwd_2d<pack_type>(team, num_cols, lid2elgp, p2d);
+        } else if (pack_alloc_property(i) == 1) {
+          set_dyn_to_zero<small_pack_type>(team);
+          team.team_barrier();
```

Comment:
Not that it matter much (I think), but the barrier here is redundant. There's no other code executed during this iteration, and there is already a guaranteed team barrier at the end of the iteration.

Reply:
Testing with OpenMP, it seems they are necessary.  The call `set_dyn_to_zero();` will iterate through the `dyn` entries differently than the call `local_remap_fwd_2d()` directly below, so I believe a `dyn` entry could be set correctly, and then set to 0. 

An option could be to iterate in `set_dyn_to_zero()` over all elements not touched in `local_remap_fwd_2d()`, but tat may be overly complicated for its benefit.

Reply:
Ah, shoes, I missed the call to `local_remap_fwd_2d`. Good point! Even if they iterated in the same way, it's dangerous to assume the same thread will access the same point in both calls.